L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+
+  typedef edm::AssociationMap<edm::OneToMany<CaloParticleCollection, l1t::HGCalTriggerCellBxCollection>> CaloToCellsMap;
+
+  bool makeCellsCollection_;
+  edm::EDGetTokenT<CaloParticleCollection> caloParticlesToken_;
+  edm::EDGetTokenT<l1t::HGCalTriggerCellBxCollection> triggerCellsToken_;
+  // // three simhits tokens, depedent on old or new geometry
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenEE_;
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEfront_;
+  // edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEback_;
+
+  // // for input with old (<= V8) geometry, sim and reco hits use different ID schemes
+  // bool needSimToRecoMapping_{false};
+
+  HGCalClusteringDummyImpl dummyClustering_;
+  HGCalShowerShape showerShape_;
+};
+
+CaloTruthCellsProducer::CaloTruthCellsProducer(edm::ParameterSet const& _config) :
+  makeCellsCollection_(_config.getParameter<bool>("makeCellsCollection")),
+  caloParticlesToken_(consumes<CaloParticleCollection>(_config.getParameter<edm::InputTag>("caloParticles"))),
+  triggerCellsToken_(consumes<l1t::HGCalTriggerCellBxCollection>(_config.getParameter<edm::InputTag>("triggerCells"))),
+  // simHitsTokenEE_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsEE"))),
+  // simHitsTokenHEfront_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEfront"))),
+  // simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEback")))
+  dummyClustering_(_config.getParameterSet("dummyClustering"))
+{
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>("HGCalBackendLayer1Processor2DClustering");
+  produces<l1t::HGCalMulticlusterBxCollection>("HGCalBackendLayer2Processor3DClustering");
+  if (makeCellsCollection_)
+    produces<l1t::HGCalTriggerCellBxCollection>("HGCalVFEProcessorSums");
+}
+
+CaloTruthCellsProducer::~CaloTruthCellsProducer()
+{
+}
+
+void
+CaloTruthCellsProducer::produce(edm::Event& _event, edm::EventSetup const& _setup)
+{
+  edm::Handle<CaloParticleCollection> caloParticlesHandle;
+  _event.getByToken(caloParticlesToken_, caloParticlesHandle);
+  auto& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  _event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto& triggerCells(*triggerCellsHandle);
+
+  // std::map<uint32_t, double> hitMap(makeHitMap(_event, _setup)); // cellId -> sim energy
+  // std::map<uint32_t, double> tcMap; // tcId -> total sim energy
+
+  dummyClustering_.eventSetup(_setup);
+  showerShape_.eventSetup(_setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  _setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto& geometry(*geometryHandle);
+
+  std::map<uint32_t, CaloParticleRef> tcToCalo;
+
+  // used later to order multiclusters
+  std::vector<CaloParticleRef> orderedCaloRefs;
```

Comment:
Could be a `std::set<CaloParticleRef>`, `Ref` has a comparator operator and the `set` will automatically order them.

Reply:
But the comparator in Ref won't give us the correct ordering here (by gen particle index of the calo particle). Plus I believe vector performs better in this use case (just storing and iterating in order).

Reply:
The Ref comparator will give the same order as the calo particle collection. I'm not sure how it differs from the `genPartIndex` you are using?

Otherwise you could use a `map<int, CaloParticleRef>`. If you just need to iterate in order and doesn't need random access, a `set` or a `map` should be better suited. Plus, the resizing of the vector is not very nice...

Reply:
For example CaloParticle [0] can have genPartIndex = 1 and CaloParticle [1] can have genPartIndex = 0.
I opted to using map<int, Ref> as you suggested. I also didn't like resizing the vectors but wrote the original code with it because the lookup would be faster.